### PR TITLE
Refresh submodule list when .hgsub changes

### LINKED ISSUE #187: Subrepos that have been removed are not converted correctly, cannot init submodules
Please see attached zip of an example mercurial repo with subrepos wherein one of the subrepos is not used in a certain branch after a certain change set.  

srt stands for "SubRepo Test".  The top level repo is srt_top, and it has two subrepos srt_suba and srt_subb.  Sorry for the naming; I apparently did not eat my Wheaties this morning. :)

For some time both subrepos are used by the top-level repo, but at change set e88d9415ea8e1a2a025b9df4c1054f2c2787cc72 srt_suba is removed from use by removing its entry in .hgsub.

I cloned out the subrepos to their own folders outsidee srt_top.  I converted the subrepos first, then the top level via the steps in README-Submodules.md

When I do the following in the git repo
`git checkout v3`
`git submodule update --init`

I receive the error:
> fatal: No url found for submodule path 'srt_suba' in .gitmodules

This also happens when using [Sourcetree](https://www.sourcetreeapp.com/) to navigate between branches.  This is the full output when attempting to checkout v3 branch:

> git -c diff.mnemonicprefix=false -c core.quotepath=false --no-optional-locks checkout v3 --progress
> M	srt_subb
> Switched to branch 'v3'
> git -c  diff.mnemonicprefix=false -c core.quotepath=false --no-optional-locks submodule update --init
> fatal: No url found for submodule path 'srt_suba' in .gitmodules
> Completed with errors, see above.

I have found some forum posts like [this](https://community.atlassian.com/t5/Sourcetree-questions/SourceTree-says-No-url-found-for-submodule-path/qaq-p/628869) and [this](https://stackoverflow.com/questions/4185365/no-submodule-mapping-found-in-gitmodule-for-a-path-thats-not-a-submodule) and doing the "git rm --cached X" will work, but only forward in time.  I tried re-writing history to add a commit that does the "git rm...." with "git rebase -i" via [this post](https://stackoverflow.com/questions/32315156/how-to-inject-a-commit-between-some-two-arbitrary-commits-in-the-past) but the interactive rebase had some merge problems in my actual repo and I'd rather not have to re-write history every changeset that removes a subrepo.

Please let me know if there is anything you need from me.  I am in charge of converting our mercurial repos to git repos because [bitbucket](https://bitbucket.org/blog/sunsetting-mercurial-support-in-bitbucket).  Right now I'm not familiar enough with the hg-fast-export tool to make the change myself and submit a pull request.  I am happy to beta test, though I suspect the attached example will be sufficient for verifying the first draft.

[srt_top.zip](https://github.com/frej/fast-export/files/4060496/srt_top.zip)
Edit: Adding 
[srt_top_submodule_mappings.txt](https://github.com/frej/fast-export/files/4073560/srt_top_submodule_mappings.txt)

### PR DESCRIPTION
**Pull request by MokhamedDakhraui**
Refresh list of submodules when `.hgsub` file changes.

close #187

### CODE REVIEW

**Comment by MokhamedDakhraui** [comment]
This is sloppy hotfix for now. At least it allows to set label `patch-exists` :smile:

File: hg-fast-export.py [hunk 1]
```diff
@@ -336,7 +336,7 @@ def get_branchname(name):
     if fn_encoding:
       filename=filename.decode(fn_encoding).encode('utf8')
     filename=strip_leading_slash(filename)
-    if filename=='.hgsubstate':
+    if filename=='.hgsubstate' or filename=='.hgsub':
```
Comment (MokhamedDakhraui) [inline hg-fast-export.py]: Use `context.substate` to check subrepos

Smth like:
```suggestion
    if (filename=='.hgsubstate' or filename=='.hgsub') and not context.substate:
```

**Comment by jdanielp** [comment]
@MokhamedDakhraui I just tested this change with the srt_top repo attached to #187.  It does not fix the issue.  This change only happens when the `.hgsub` or `.hgsubstate` file is removed, but your comment above says it should happen when the file is changed.  There is no such loop for changed files in the `export_commit` function, only for removed files.

**Comment by MokhamedDakhraui** [comment]
Before make PR I tested it on `srt_top` repo from #187. On v3 command git submodules shows only `srt_subb`.
Submodules list should be refreshed when .hgsub-ish files added/modified here:
https://github.com/frej/fast-export/blob/4e8c6bbc9747bffc6abbb62b7f6a725d82989b44/hg-fast-export.py#L195

Can you apply resulted git repo and point to commit after which problem occurs.
I also replay test on my end latter

**Comment by MokhamedDakhraui** [comment]
I'm just contributor so I the final decision will make @frej 

If just for opinion, then it seems that your patch do the same as `refresh_submodules` but more selectively (not just remove all submodules and then readd relevant). So I suggest you to refactor `refresh_submodules` to add/remove submodules in more optimal way in new PR

**Comment by jdanielp** [comment]
@MokhamedDakhraui I tested your fix on my work repo and it looks good.  I am able to checkout the various branches without issue.  

For the record: I did not include the suggested change for `and not context.substate` on line 339

**Comment by alexey-lcard** [comment]
Fix works great on my repos too. Many thanks!

File: hg-fast-export.py [hunk 2]
```diff
@@ -192,7 +192,7 @@ def export_file_contents(ctx,manifest,files,hgtags,encoding='',plugins={}):
   count=0
   max=len(files)
   for file in files:
-    if file==".hgsubstate":
+    if file=='.hgsub' or file='.hgsubstate':
```
Comment (frej) [inline hg-fast-export.py]:
```Python
file='.hgsubstate'
```
Isn't valid python. (`==` is probably what you intended, and with that this patch seems to fix #87)

**Comment by frej** [comment]
@MokhamedDakhraui : I really appreciate that you support your previous contributions. I'm much more likely to accept patches when I know that the author will provide support if bugs turn up later on. Thanks for the help!

File: hg-fast-export.py (hunk 2 above)
Comment (MokhamedDakhraui) [reply hg-fast-export.py]: Fixed

**Comment by MokhamedDakhraui** [comment]
Optimized a little to refresh submodules only once after crossing over either `.hgsub` or `.hgsubstate`. I guess it's ready for review

**Comment by jdanielp** [comment]
@frej I'm fine with what you said about using @MokhamedDakhraui 's fix.  If you want me to test the fix and review the changes as a second reviewer I can do that on Monday.

I submited the patch as a zip since I cannot submit .diff or .patch files as attachments.  Should I have initiated my own pull request with my patch?  I did not do so because it would have split the conversation onto two different threads if I submitted as a new PR.  Or maybe that's what's supposed to happen.  I'm a newbie when it comes to working with open source projects.

I plan to work on #188 early next week.  If I get to a fix first I'll be sure to use a PR there.

**Comment by frej** [comment]
@jdanielp, I'll take you up on the offer to run the patch on your internal repo, let me know how it goes.

**Comment by jdanielp** [comment]
@frej @MokhamedDakhraui @alexey-lcard 

I applied the patch from https://github.com/frej/fast-export/pull/189.patch to my working copy and converted my work repo.  Everything looks good when I switch between branches in sourcetree and in git bash on windows.  I'd say we're good to go.

**Pull request merged**